### PR TITLE
Add JPEG size hint to ImageMagick command

### LINKED ISSUE #3023: IM thumb engine fails at processing large image files
Kirby 3.4.3 (maybe 3.4.x) reintroduced the issue #1788 and reported by @silllli

### PR DESCRIPTION
**Pull request by silllli**
## Describe the PR

One step in enhancing Kirby’s thumbnail creation, especially with large images in resource-restricted server environments. It adds a “size hint” (see [ImageMagick docs](https://imagemagick.org/script/defines.php)) when calling ImageMagick with a JPEG file which can reduce the CPU and memory usage drastically—most notably when the output file is a lot smaller than the original.

### Enhancements
* JPEG size hint added to the ImageMagick thumb driver to increase performance with large source images

## Breaking changes

None

## Related issues/ideas

- Partly fixes #3023 

## Ready?

- [ ] ~Unit tests for fixed bug/feature~ (hard if not impossible to test because it's an internal working of IM)
- [x] In-code documentation (wherever needed)
- [x] CI checks pass

## When merging

- [ ] ~Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)~
- [x] Add changes to release notes draft in Notion

### CODE REVIEW

File: src/Image/Darkroom/ImageMagick.php
Comment (lukasbestle) [inline src/Image/Darkroom/ImageMagick.php]:
```suggestion
                $command .= ' -define ' . escapeshellarg(sprintf('jpeg:size=%dx%d', $options['width'], $options['height']));
```

**Review by lukasbestle — COMMENTED** [review]
Thank you for your contribution, the change makes a lot of sense. 👍

Unfortunately I saw your PR after I created #4109 from your code comment. The two PRs conflict due to changes in the same file, so I suggest we first wait until the other PR is merged. I'll then rebase this PR.

I have two suggestions for your code:

- To increase performance, the fact that the image is being scaled down could be stored in the `$options` array in `Image\Darkroom::preprocess()` as this method already has the `$image` object at hand. There could be new float `$option['scaleWidth']` and `$option['scaleHeight']` options for this (defined as the scale ratio between the thumb and the original width and height). These options should be set to `null` by default in the `defaults()` method.
- The new command arguments should use `escapeshellarg()`, see below.

**Comment by silllli** [comment]
> Unfortunately I saw your PR after I created https://github.com/getkirby/kirby/pull/4109 from your code comment. The two PRs conflict due to changes in the same file, so I suggest we first wait until the other PR is merged. I'll then rebase this PR.

No problem, could you rebase the PR now?

Also: as I have suggested in the issue, I have some more resource optimization ideas. Should I add them in this PR or should I stick to the size hint with this one and create another one for the memory `-limit` option?

File: src/Image/Darkroom/ImageMagick.php
Comment (silllli) [reply src/Image/Darkroom/ImageMagick.php]: > Unfortunately I saw your PR after I created https://github.com/getkirby/kirby/pull/4109 from your code comment. The two PRs conflict due to changes in the same file, so I suggest we first wait until the other PR is merged. I'll then rebase this PR.

No problem, could you rebase the PR now?

Also: as I have suggested in the issue, I have some more resource optimization ideas. Should I add them in this PR or should I stick to the size hint with this one and create another one for the memory `-limit` option?

Comment (distantnative) [reply src/Image/Darkroom/ImageMagick.php]: @silllli I don't think we can rebase this branch cause it's based on your fork. You could rebase it from `develop` branch or I copy your branch etc. and rebase it - but this would then require closing this PR and open a new one from the copied branch.

And I think separate PR would be helpful. Always easier to review one thing at a time.

Comment (distantnative) [reply src/Image/Darkroom/ImageMagick.php]: @silllli which one do you prefer (you rebasing your PR or me opening a new PR)?

Comment (silllli) [reply src/Image/Darkroom/ImageMagick.php]: I will rebase it 👍

Comment (silllli) [reply src/Image/Darkroom/ImageMagick.php]: I’ve rebased the commit and added @lukasbestle’s suggestions.

(By the way I noticed that `preprocess()` is called twice during thumb generation. Not sure if that is intended and / or needed or not.)

Comment (silllli) [reply src/Image/Darkroom/ImageMagick.php]: Oops, I’m not a git genius. All these commits that happened in the meantime shouldn’t be part of the PR I guess. I will try to fix it.

Comment (silllli) [reply src/Image/Darkroom/ImageMagick.php]: Seems to look good now, after some git black magic…

**Review by lukasbestle — DISMISSED** [review]
Thank you very much for your work and sorry for the huge delay. In the last weeks I was busy with other things and only now came around to review your PR again.

I like the implementation and think it's already a great improvement.

Regarding `preprocess()` being called multiple times: One time it is called from the `thumb` component to determine the options to build the filename, and then the driver itself also calls it. With the current setup there is no way around that as the options are not cached inside `$darkroom`. But it could be improved in a later step.

**Comment by lukasbestle** [comment]
Oh, the unit tests fail. I will fix them tomorrow evening.

**Comment by mrflix** [comment]
@silllli I'm a bit late with commenting here: the ImageMagick docs suggest that the `jpg:size` dimensions should be double the size of the final image to prevent antialiasing effects:

> Note that this modifier causes the JPEG library to skip the reading of whole columns and rows of pixels. As such it will produce effects much like the [Sampling Resize Operator](https://imagemagick.org/Usage/resize/#sample), including its strong [Aliasing Artefacts](https://imagemagick.org/Usage/filter/#aliasing).
**Because of this is it recommended that you specify at least double the final 'resize' of the image, to avoid this problem, just as shown in the above example.**

https://imagemagick.org/Usage/formats/#jpg_read

**Comment by silllli** [comment]
> @silllli I'm a bit late with commenting here: the ImageMagick docs suggest that the `jpg:size` dimensions should be double the size of the final image to prevent antialiasing effects:
> 
> > Note that this modifier causes the JPEG library to skip the reading of whole columns and rows of pixels. As such it will produce effects much like the [Sampling Resize Operator](https://imagemagick.org/Usage/resize/#sample), including its strong [Aliasing Artefacts](https://imagemagick.org/Usage/filter/#aliasing).
> > **Because of this is it recommended that you specify at least double the final 'resize' of the image, to avoid this problem, just as shown in the above example.**
> 
> https://imagemagick.org/Usage/formats/#jpg_read

I think I remember reading that and tested it myself with some images where I didn’t notice the effect so I dismissed it. I also remember the IM docs as seeming partially outdated so I figured it might not be a problem anymore. I had not seen their examples where the effect is visible. Sorry!

**Comment by mrflix** [comment]
Don't say sorry! You might as well be right. I think the examples are examples of the effect itself (sampling & antialiasing), not results of the `jpeg:size` setting. I tried to understand it's [ImageMagick code](https://github.com/ImageMagick/ImageMagick/blob/acaffeb8e496ab23f6b56f3411ff8e69c7ff2fe3/coders/jpeg.c#L1201), but I don't fully understand it. In the docs it says:
> Remember this is only a hint as the size of the image wanted, you are not guaranteed to get this size, just something close-to but larger than that size. **Typically you will get something that is between this size and one twice that big, while preserving the images aspect ratio.** 

[Source](https://imagemagick.org/Usage/formats/#jpg_read)

This might have to do with libjpeg, which is used under the hood, only supporting scaling down the image by certain factors:
```
unsigned int scale_num, scale_denom
        Scale the image by the fraction scale_num/scale_denom.  Default is
        1/1, or no scaling.  Currently, the only supported scaling ratios
        are M/8 with all M from 1 to 16, or any reduced fraction thereof (such
        as 1/2, 3/4, etc.)  (The library design allows for arbitrary
        scaling ratios but this is not likely to be implemented any time soon.)
        Smaller scaling ratios permit significantly faster decoding since
        fewer pixels need be processed and a simpler IDCT method can be used.
```
[Source](https://github.com/winlibs/libjpeg/blob/6904e8cd02785f4faa500714d4d1e150a3497200/libjpeg.txt#L1224)

[Here](https://stackoverflow.com/a/9878900/731172) it says that that's true for libjpeg 8c+ and older libjpeg versions only support coarser downscaling.

My example image in #5855 has the dimensions 3543x2362. I want to crop it to 512x683. With the setting `-define 'jpeg:size=512x683'`, ImageMagick calculates a scaling factor of 0.2891617273. Which is the longer side of the crop divided by the shorter side of the image 683 / 2362. So libjpeg will use a factor of 3/8 = 0.375. Maybe that's good enough.

**Pull request merged**
